builder/amazon/common/step_create_ssm_tunnel.go

```diff
@@ -0,0 +1,135 @@
+	"github.com/hashicorp/packer/packer"
+)
+
+type StepCreateSSMTunnel struct {
+	AWSSession      *session.Session
+	DstPort         int
+	SSMAgentEnabled bool
+	instanceId      string
+	ssmSession      *ssm.StartSessionOutput
+}
+
+func (s *StepCreateSSMTunnel) Run(ctx context.Context, state multistep.StateBag) multistep.StepAction {
+	if !s.SSMAgentEnabled {
+		return multistep.ActionContinue
+	}
+
+	ui := state.Get("ui").(packer.Ui)
+	// Find an available TCP port for our HTTP server
+	l, err := net.ListenRangeConfig{
+		Min:     8000,
+		Max:     9000,
+		Addr:    "0.0.0.0",
+		Network: "tcp",
+	}.Listen(ctx)
+	if err != nil {
+		err := fmt.Errorf("error finding an available port to initiate a session tunnel: %s", err)
+		state.Put("error", err)
+		ui.Error(err.Error())
+		return multistep.ActionHalt
+	}
+
+	dst, src := strconv.Itoa(s.DstPort), strconv.Itoa(l.Port)
+	params := map[string][]*string{
+		"portNumber":      []*string{aws.String(dst)},
+		"localPortNumber": []*string{aws.String(src)},
+	}
+	l.Close()
+
+	instance, ok := state.Get("instance").(*ec2.Instance)
+	if !ok {
+		err := fmt.Errorf("error encountered in obtaining target instance id for SSM tunnel")
+		ui.Error(err.Error())
+		state.Put("error", err)
+		return multistep.ActionHalt
+	}
+
+	s.instanceId = aws.StringValue(instance.InstanceId)
+	ssmconn := ssm.New(s.AWSSession)
+	input := ssm.StartSessionInput{
+		DocumentName: aws.String("AWS-StartPortForwardingSession"),
+		Parameters:   params,
+		Target:       aws.String(s.instanceId),
+	}
+
+	ui.Message(fmt.Sprintf("Starting PortForwarding session to instance %q on local port %q to remote port %q", s.instanceId, src, dst))
+	var output *ssm.StartSessionOutput
+	err = retry.Config{
+		ShouldRetry: func(err error) bool { return isAWSErr(err, "TargetNotConnected", "") },
+		RetryDelay:  (&retry.Backoff{InitialBackoff: 200 * time.Millisecond, MaxBackoff: 60 * time.Second, Multiplier: 2}).Linear,
+	}.Run(ctx, func(ctx context.Context) error {
+		output, err = ssmconn.StartSessionWithContext(ctx, &input)
+		return err
+	})
+
+	if err != nil {
+		err = fmt.Errorf("error encountered in starting session for instance %q: %s", s.instanceId, err)
+		ui.Error(err.Error())
+		state.Put("error", err)
+		return multistep.ActionHalt
+	}
+	s.ssmSession = output
+
+	// AWS session-manager-plugin requires a valid session be passed in JSON
+	sessionDetails, err := json.Marshal(s.ssmSession)
+	if err != nil {
+		ui.Error(err.Error())
+		state.Put("error encountered in reading session details", err)
+		return multistep.ActionHalt
+	}
+
+	sessionParameters, err := json.Marshal(input)
+	if err != nil {
+		ui.Error(err.Error())
+		state.Put("error encountered in reading session parameter details", err)
+		return multistep.ActionHalt
+	}
+
+	driver := SSMDriver{Ui: ui}
+	// sessionDetails, region, "StartSession", profile, paramJson, endpoint
+	region := aws.StringValue(s.AWSSession.Config.Region)
+	// how to best get Profile name
```

Comment:
Is this a question reminder? 🤔

Reply:
A little bit of both :laughing: But in my testing it seems that no actual profile information is needed for the command to work. I think once I have more details from users I can determine if it is needed. Do you know of a way to get to this information through an active session?